captum/attr/_utils/lrp_rules.py

```diff
@@ -0,0 +1,148 @@
+            self.relevance_output = grad.data
+            return relevance
+
+        return _backward_hook_output
+
+    def forward_hook_weights(self, module, inputs, outputs):
+        """Save initial activations a_j before modules are changed"""
+        module.activations = tuple(input.data for input in inputs)
+        self._manipulate_weights(module, inputs, outputs)
+
+    @abstractmethod
+    def _manipulate_weights(self, module, inputs, outputs):
+        raise NotImplementedError
+
+    def forward_pre_hook_activations(self, module, inputs):
+        """Pass initial activations to graph generation pass"""
+        for input, activation in zip(inputs, module.activations):
+            input.data = activation
+        return inputs
+
+
+class EpsilonRule(PropagationRule):
+    """
+    Rule for relevance propagation using a small value of epsilon
+    to avoid numerical instabilities and remove noise.
+
+    Use for middle layers.
+
+    Args:
+        epsilon (integer, float): Value by which is added to the
+        discriminator during propagation.
+    """
+
+    def __init__(self, epsilon=1e-9):
+        self.STABILITY_FACTOR = epsilon
+
+    def _manipulate_weights(self, module, inputs, outputs):
+        pass
+
+
+class GammaRule(PropagationRule):
+    """
+    Gamma rule for relevance propagation, gives more importance to
+    positive relevance.
+
+    Use for lower layers.
+
+    Args:
+        gamma (float): The gamma parameter determines by how much
+        the positive relevance is increased.
+    """
+
+    def __init__(self, gamma=0.25, set_bias_to_zero=False):
+        self.gamma = gamma
+        self.set_bias_to_zero = set_bias_to_zero
+
+    def _manipulate_weights(self, module, inputs, outputs):
+        if hasattr(module, "weight"):
+            module.weight.data = (
+                module.weight.data + self.gamma * module.weight.data.clamp(min=0)
+            )
+        if self.set_bias_to_zero and hasattr(module, "bias"):
+            if module.bias is not None:
+                module.bias.data = torch.zeros_like(module.bias.data)
+
+
+class Alpha1_Beta0_Rule(PropagationRule):
+    """
+    Alpha1_Beta0 rule for relevance backpropagation, also known
+    as Deep-Taylor. Only positive relevance is propagated, resulting
+    in stable results, therefore recommended as the initial choice.
+
+    Warning: Does not work for BatchNorm modules because weight and bias
+    are defined differently.
+
+    Use for lower layers.
+    """
+
+    def __init__(self, set_bias_to_zero=False):
+        self.set_bias_to_zero = set_bias_to_zero
+
+    def _manipulate_weights(self, module, inputs, outputs):
+        if hasattr(module, "weight"):
+            module.weight.data = module.weight.data.clamp(min=0)
+        if self.set_bias_to_zero and hasattr(module, "bias"):
+            if module.bias is not None:
+                module.bias.data = torch.zeros_like(module.bias.data)
+
+
+class IdentityRule(EpsilonRule):
```

Comment:
I think this does not work as intended due to the order in which hooks are invoked on intermediate tensors, in particular if this is attached on layer l, the _backward_hook_input on any input tensor of layer l is executed after a putative _backward_hook_output of layer l-1 that is attached to the same intermediate tensor.
It clears the test_lrp_identity due to the relu between the layers as the relu gets a module backward hook attached and therefore does not attach hooks on the intermediate results.

Reply:
@rGure, do you mean that this works in the use case described in `test_lrp_identity ` but will fail if we have two linear layers following each other ?